### PR TITLE
660 log transfer operations

### LINKED ISSUE #660: Log transfer operations
Using the same logging that is done in services for shipments/requisition, add logs for insert, update, deletes.
We would like to show in UI that it was a system user that did the operation.

### A - Created on central

Similar to how we do name for inventory adjustment. This use will need to be created on central server when new data files is created or on migration. 

**pros**
* Consistent with how we do inventory adjustments etc

**cons**
* It would have to be a special user that no one can log into and that is not editable 
* Changes would need to be done on central

### B - Hard coded system user

Although for (A) some hard coding is done (the code of system user), with B user is completely hard coded. In service layer, when `code` or `id` of user is `system` then return a hard coded UserRow.

**pros**
* No cons of A
* Fast to implement and allows to transition to A

**cons**
* No pros of A

`consideration` -> when logging user in, and user is not found on central, when we confirm user/password matches locally we should filter out users with no password (which system user will have)

## System user decision

@andreievg votes for B

### Extra

This is from a placeholder point from the epic: 

Should system user be used/displayed for transfer creation/edit ? And at the same time we want to make sure we log operations like we do in services https://github.com/openmsupply/open-msupply/pull/581#discussion_r971429435 https://github.com/openmsupply/open-msupply/pull/581#discussion_r971430336

### PR DESCRIPTION
**Pull request by roxy-dao**
Closes #660 

- Went with option B...
- Transfers for invoices and requisitions now logged into the log table

### CODE REVIEW

File: server/service/src/processors/transfer/requisition/create_response_requisition.rs [hunk 1]
```diff
@@ -114,6 +116,16 @@ fn generate_response_requisition(
         comment: None,
     };
 
+    system_log_entry(
+        connection,
+        match result.status.clone() {
+            RequisitionRowStatus::New => LogType::RequisitionCreated,
+            _ => LogType::RequisitionCreated,
+        },
```
Comment (mark-prins) [inline server/service/src/processors/transfer/requisition/create_response_requisition.rs:124]: Should the default arm be doing something else?
If not - remove the match?

File: server/service/src/processors/transfer/shipment/create_inbound_shipment.rs [hunk 2]
```diff
@@ -145,5 +145,16 @@ fn generate_inbound_shipment(
         verified_datetime: None,
     };
 
+    system_log_entry(
+        connection,
+        match result.status.clone() {
+            InvoiceRowStatus::Picked => LogType::InvoiceStatusPicked,
+            InvoiceRowStatus::Shipped => LogType::InvoiceStatusShipped,
+            _ => LogType::InvoiceCreated,
```
Comment (mark-prins) [inline server/service/src/processors/transfer/shipment/create_inbound_shipment.rs:153]: given that the fn is `generate_inbound_shipment` I wonder if this should always be `InvoiceCreated`?
I think I understand - in which case, yeah. it might be helpful to have the status of the created invoice.

Side question: is the non-exhaustive match ok? If we later add another value to the enum, then wouldn't it be better for the compiler to tell us that we need to implement a mapping? As it is, the compiler won't complain and the status will be `InvoiceCreated` which might not be correct / accurate..

File: server/service/src/system_user.rs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+use repository::{RepositoryError, StorageConnection, UserAccountRow, UserAccountRowRepository};
+use util::uuid::uuid;
+
+pub fn system_user(connection: &StorageConnection) -> Result<UserAccountRow, RepositoryError> {
+    let system_user = UserAccountRow {
+        id: uuid(),
```
Comment (mark-prins) [inline server/service/src/system_user.rs:6]: I wonder if this might be better as a static id - generating a new UUID on every access will require searching for the `System` username: it isn't possible to search for the system user by ID now

**Review by mark-prins — COMMENTED** [review]
Overall, that's awesome! Just what we need. 
One concern: I think a status needs updating?

I'm conflicted about whether the status should be 'created' or reflect the status of the other half of process, and wonder about using an exhaustive match rather than the catch-all arm.

File: server/service/src/system_user.rs (hunk 3 above)
Comment (roxy-dao) [reply server/service/src/system_user.rs:6]: I had the same thought... but the Static ID gave a 'NOT UNIQUE' error when trying to transfer requisitions 🤔

File: server/service/src/processors/transfer/requisition/create_response_requisition.rs (hunk 1 above)
Comment (roxy-dao) [reply server/service/src/processors/transfer/requisition/create_response_requisition.rs:124]: All the patterns need to be handled, so can't remove the wildcard... but can change it to a different LogType if you want?

File: server/service/src/processors/transfer/shipment/create_inbound_shipment.rs (hunk 2 above)
Comment (roxy-dao) [reply server/service/src/processors/transfer/shipment/create_inbound_shipment.rs:153]: Outbound transfers should only be possible when the status is 'Picked' or 'Shipped', so the other statuses shouldn't trigger the transfer 🤞

File: server/service/src/processors/transfer/requisition/create_response_requisition.rs [hunk 4]
```diff
@@ -114,6 +116,16 @@ fn generate_response_requisition(
         comment: None,
     };
 
+    system_log_entry(
```
Comment (andreievg) [inline server/service/src/processors/transfer/requisition/create_response_requisition.rs:119]: Can the logs be inside processor please. (they are less `hidden` away)

File: server/service/src/system_user.rs (hunk 3 above)
Comment (andreievg) [reply server/service/src/system_user.rs:6]: We don't need to insert the row at all, sorry I should have been more specific, what i was meaning, similar to special name:
* Make a new constant here: https://github.com/openmsupply/open-msupply/blob/develop/server/util/src/constants.rs, something like 'SYSTEM_USER_ID =  "omsupply_system_user`
* Use that constant in processor logs, ooh, we can make a new 'system user context method' and use here: https://github.com/openmsupply/open-msupply/blob/7043d27cb03796af38dfbc5a7ab57464ddebec92/server/service/src/processors/transfer/requisition/mod.rs#L68
* Whenever we query for user we should intercept user id with SYSTEM_USER_ID and return the system user, i thought this should be done in `service` layer but just realised that we don't yet use service layer in user loader : /. So for now, although it's a little bit hacky, add that user to the loader result if `keys` includes SYSTEM_USER_ID.

**Comment by andreievg** [comment]
One question remains, what user do we associate with the created requisition/invoice (it's none atm), can be system user or the other half of transfer user.

File: server/service/src/processors/transfer/requisition/create_response_requisition.rs (hunk 1 above)
Comment (andreievg) [reply server/service/src/processors/transfer/requisition/create_response_requisition.rs:124]: I think Mark is asking, why is there a match ?

**Comment by andreievg** [comment]
Was just looking at log types, and realised we need to add two logs records in some cases:

```
    InvoiceCreated,
    InvoiceDeleted,
    InvoiceStatusAllocated,
    InvoiceStatusPicked,
    InvoiceStatusShipped,
    InvoiceStatusDelivered,
    InvoiceStatusVerified,
    StocktakeCreated,
    StocktakeDeleted,
    StocktakeStatusFinalised,
    RequisitionCreated,
    RequisitionDeleted,
    RequisitionStatusSent,
    RequisitionStatusFinalised,
```

* `create_response_requisition` -> `RequisitionCreated`
* `update_request_requisition_status` -> `RequisitionStatusFinalised`
* `create_inbound_shipment` -> `InvoiceCreated` + `InvoiceStatus{status}`
* `delete_inbound_shipment` -> `InvoiceDeleted`
* `update_inbound_shipment` -> `InvoiceStatus{status}`
* `update_outbound_shipmen_status` -> `InvoiceStatus{status}`

File: server/service/src/processors/transfer/requisition/create_response_requisition.rs (hunk 1 above)
Comment (mark-prins) [reply server/service/src/processors/transfer/requisition/create_response_requisition.rs:124]: yep. the `LogType::RequisitionCreated` is returned if it is `RequisitionRowStatus::New` 
If it isn't `LogType::RequisitionCreated` then the return is also `RequisitionRowStatus::New` 
there is no need for the match - you can just set it to `LogType::RequisitionCreated` 😉 

// perhaps a comment to explain why the type is always `RequisitionCreated` though?

I was wondering though, if the wildcard was supposed to be returning a different value?

Comment (roxy-dao) [reply server/service/src/processors/transfer/requisition/create_response_requisition.rs:124]: 🤦

File: server/service/src/system_user.rs (hunk 3 above)
Comment (roxy-dao) [reply server/service/src/system_user.rs]: [8563099](https://github.com/openmsupply/open-msupply/pull/666/commits/8563099e0beb5c369b20e12dd9e36bef7fb9f478)

File: server/service/src/processors/transfer/requisition/create_response_requisition.rs (hunk 4 above)
Comment (roxy-dao) [reply server/service/src/processors/transfer/requisition/create_response_requisition.rs]: [90faf29](https://github.com/openmsupply/open-msupply/pull/666/commits/90faf29237d024125225ff80f11c21841b978f1f)

**Comment by roxy-dao** [comment]
> a80f579

File: server/service/src/system_user.rs [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    let system_user = UserAccountRow {
+        id: SYSTEM_USER_ID.to_string(),
+        username: SYSTEM_USER_ID.to_string(),
+        hashed_password: "".to_string(),
+        email: None,
+    };
+
+    let connection = service_provider.connection()?;
+    let user_repository = UserAccountRowRepository::new(&connection);
+    match user_repository.find_one_by_id(&system_user.id)? {
```
Comment (andreievg) [inline server/service/src/system_user.rs:16]: :nit. if ....is_some()

File: server/service/src/processors/transfer/requisition/update_request_requisition_status.rs [hunk 6]
```diff
@@ -66,6 +68,13 @@ impl RequisitionTransferProcessor for UpdateRequestRequstionStatusProcessor {
 
         RequisitionRowRepository::new(connection).upsert_one(&updated_request_requisition)?;
 
+        system_log_entry(
+            connection,
+            LogType::RequisitionStatusFinalised,
+            Some(request_requisition.requisition_row.store_id.clone()),
```
Comment (andreievg) [inline server/service/src/processors/transfer/requisition/update_request_requisition_status.rs:74]: log should be for `updated_request_requisition`

File: server/service/src/processors/transfer/shipment/create_inbound_shipment.rs [hunk 7]
```diff
@@ -78,6 +78,26 @@ impl ShipmentTransferProcessor for CreateInboundShipmentProcessor {
 
         InvoiceRowRepository::new(connection).upsert_one(&new_inbound_shipment)?;
 
+        if outbound_shipment.invoice_row.status.clone() == InvoiceRowStatus::Picked {
```
Comment (andreievg) [inline server/service/src/processors/transfer/shipment/create_inbound_shipment.rs:81]: This check is not needed, this action will only fire once:
https://github.com/openmsupply/open-msupply/blob/88774b26d841ece222fba472fffb1e5980ce9fd8/server/service/src/processors/transfer/shipment/create_inbound_shipment.rs#L33-L34

File: server/service/src/processors/transfer/shipment/create_inbound_shipment.rs [hunk 8]
```diff
@@ -78,6 +78,26 @@ impl ShipmentTransferProcessor for CreateInboundShipmentProcessor {
...
+            system_log_entry(
+                connection,
+                LogType::InvoiceCreated,
+                Some(outbound_shipment.invoice_row.store_id.clone()),
```
Comment (andreievg) [inline server/service/src/processors/transfer/shipment/create_inbound_shipment.rs:85]: new_inbound_shipment

File: server/service/src/processors/transfer/shipment/create_inbound_shipment.rs [hunk 9]
```diff
@@ -78,6 +78,26 @@ impl ShipmentTransferProcessor for CreateInboundShipmentProcessor {
...
+                Some(new_inbound_shipment.id.clone()),
+            )?;
+        }
+
+        system_log_entry(
+            connection,
+            match outbound_shipment.invoice_row.status.clone() {
+                InvoiceRowStatus::Picked => LogType::InvoiceStatusPicked,
```
Comment (andreievg) [inline server/service/src/processors/transfer/shipment/create_inbound_shipment.rs:93]: I wonder if we can have just one LogType::from_invoice_status method that we use for all invoice update logs. 
I see that in service update method for shipments we match on input.status*, I think the logic there should be, did status change ? if yes lets log the new status (from invoice.status). I think it's worth updating this in this PR.

*input.status (does this mean if a comment for example is updated, we log that status was also changed ?),

File: server/service/src/processors/transfer/shipment/delete_inbound_shipment.rs [hunk 10]
```diff
@@ -69,6 +69,18 @@ impl ShipmentTransferProcessor for DeleteInboundShipmentProcessor {
         // 6.
         InvoiceRowRepository::new(connection).delete(deleted_inbound_shipment_id)?;
 
+        let store_id = match linked_shipment {
```
Comment (andreievg) [inline server/service/src/processors/transfer/shipment/delete_inbound_shipment.rs:72]: to make it slightly cleaner, i would do this:
(above)
`let deleted_inbound_shipment_id = &inbound_shipment.invoice_row.id`
`let deleted_inboud_shipment = inbound_shipment.invoice_row`

`InvoiceRowRepository::new(connection).delete(&deleted_inbound_shipment.id)?;`

 then use deleted_inbound_shipment for store_id etc..

File: server/service/src/processors/transfer/shipment/update_inbound_shipment.rs [hunk 11]
```diff
@@ -88,6 +88,17 @@ impl ShipmentTransferProcessor for UpdateInboundShipmentProcessor {
...
         InvoiceRowRepository::new(connection).upsert_one(&updated_inbound_shipment)?;
 
+        system_log_entry(
+            connection,
+            match inbound_shipment.invoice_row.status.clone() {
+                InvoiceRowStatus::Delivered => LogType::InvoiceStatusDelivered,
+                InvoiceRowStatus::Verified => LogType::InvoiceStatusVerified,
+                _ => LogType::InvoiceCreated,
+            },
+            Some(outbound_shipment.invoice_row.store_id.clone()),
```
Comment (andreievg) [inline server/service/src/processors/transfer/shipment/update_inbound_shipment.rs:98]: inbound_shipment

File: server/service/src/log.rs [hunk 12]
```diff
@@ -54,3 +55,21 @@ pub fn log_entry(
 
     Ok(LogRowRepository::new(&ctx.connection).insert_one(log)?)
 }
+
+pub fn system_log_entry(
```
Comment (andreievg) [inline server/service/src/log.rs:59]: I think both store_id and record_id in current cases are not optional ?

File: server/service/src/processors/transfer/shipment/update_outbound_shipment_status.rs [hunk 13]
```diff
@@ -71,6 +71,17 @@ impl ShipmentTransferProcessor for UpdateOutboundShipmentStatusProcessor {
...
         InvoiceRowRepository::new(connection).upsert_one(&updated_outbound_shipment)?;
 
+        system_log_entry(
+            connection,
+            match inbound_shipment.invoice_row.status.clone() {
+                InvoiceRowStatus::Delivered => LogType::InvoiceStatusDelivered,
+                InvoiceRowStatus::Verified => LogType::InvoiceStatusVerified,
+                _ => LogType::InvoiceCreated,
+            },
+            Some(inbound_shipment.invoice_row.store_id.clone()),
```
Comment (andreievg) [inline server/service/src/processors/transfer/shipment/update_outbound_shipment_status.rs:81]: outbound_shipment

File: server/service/src/processors/transfer/shipment/delete_inbound_shipment.rs (hunk 10 above)
Comment (roxy-dao) [reply server/service/src/processors/transfer/shipment/delete_inbound_shipment.rs:72]: Would the store_id not belong to the outbound shipment though since it'll be deleted at picked? 🤔

File: server/service/src/processors/transfer/shipment/create_inbound_shipment.rs (hunk 9 above)
Comment (roxy-dao) [reply server/service/src/processors/transfer/shipment/create_inbound_shipment.rs]: [2fb3de6](https://github.com/openmsupply/open-msupply/pull/666/commits/2fb3de6def14108b24c3a8d6045947d65b4d4401)

File: server/service/src/system_user.rs (hunk 5 above)
Comment (roxy-dao) [reply server/service/src/system_user.rs]: [0280520](https://github.com/openmsupply/open-msupply/pull/666/commits/028052068814657307c363cb76634735f38b733d)

**Comment by andreievg** [comment]
@roxy-dao

What I meant by 
```
 I wonder if we can have just one LogType::from_invoice_status method that we use for all invoice update logs.
I see that in service update method for shipments we match on input.status*, I think the logic there should be, did status change ? if yes lets log the new status (from invoice.status). I think it's worth updating this in this PR.

*input.status (does this mean if a comment for example is updated, we log that status was also changed ?),
```

* create LogType::from_invoice_status method to be used in transfer logging and in existing logging
* current logging code allows duplicate status log, api consumer can send the same status (quite often the consumer when updating just say a comment will post the whole input, including existing, unchanged status), I've added the [test](https://github.com/openmsupply/open-msupply/compare/660-Log-transfer-operations...660-tests-for-duplicates): 
* so basically we currently map from ShipmentStatusInput like this:

https://github.com/openmsupply/open-msupply/blob/5778345e003ccc9b00a5530cd0ae07a9e2d89c7f/server/service/src/invoice/outbound_shipment/update/mod.rs#L97-L108

Would be good to map from InvoiceRowStatus if it was changed,

File: server/service/src/invoice/inbound_shipment/update/mod.rs [hunk 14]
```diff
@@ -43,6 +43,7 @@ pub fn update_inbound_shipment(
         .connection
         .transaction_sync(|connection| {
             let (invoice, other_party) = validate(connection, &ctx.store_id, &patch)?;
+            let status_changed = check_status_change(&invoice, patch.full_status());
```
Comment (roxy-dao) [inline server/service/src/invoice/inbound_shipment/update/mod.rs:46]: Log wouldn't insert when checked in validation... 🤔

**Comment by andreievg** [comment]
@roxy-dao thanks for the changes 🙏 

I did end up adjusting things to show what I meant about the suggestion:  https://github.com/openmsupply/open-msupply/compare/660-Log-transfer-operations...660-further-refactor

* Simplify log_entry slightly by passing record_is as ref
* Move all status_changed to validate
* One log_type_from_invoice_status method (the origin of my suggestion really), which also meant i removed system log helper that does the conversion
* Adjusted some logging in transfers to log status for record being edited

There was some unexpected changes, in stocktake, that spiralled a little bit, for some reason there was an option to update stocktake to `new` status, similar to requisitions, status for stocktake could only be updatable to finalised. I've updated back and after re generating front end type from back end schema, fixed compilation errors on front end.

**Review by andreievg — APPROVED** [review]
Approve pending changes from latest suggestion merge into this branch

**Comment by andreievg** [comment]
going to resolve conflicts in 660-further-refactor

**Comment by andreievg** [comment]
@roxy-dao i resolved conflicts on `660-further-refactor-update-from-develop` branch, couldn't do it on this branch since there are some changes in `660-further-refactor` that would require re-generating front end types from API, and couldn't do it on `660-further-refactor` as there will be too much diff between current state of this branch and `660-further-refactor`. You will need to merge `660-further-refactor` into this branch first, then merge `660-further-refactor-update-from-develop` into this branch, and then it's good to go : D
